velox/exec/Merge.cpp

```diff
@@ -165,60 +200,186 @@ RowVectorPtr Merge::getOutput() {
     return data;
   }
 
-  if (!output_) {
-    output_ = BaseVector::create<RowVector>(
-        outputType_, outputBatchSize_, operatorCtx_->pool());
-    for (auto& child : output_->children()) {
-      child->resize(outputBatchSize_);
+  VELOX_CHECK_NOT_NULL(sourceMerger_);
+  bool lastRun = false;
+  output_ = sourceMerger_->getOutput(
+      outputBatchSize_, sourceBlockingFutures_, lastRun);
+
+  // mergeBuffer spill maybe finished.
+  if (needSpill()) {
+    spill();
+    if (lastRun) {
+      finishSpill();
     }
+    return nullptr;
   }
 
-  for (;;) {
-    auto stream = treeOfLosers_->next();
+  VELOX_CHECK_EQ(numStartedSources_, sources_.size());
+  // Create spillMerger_ exactly once if spill has happened.
+  if (!spillReadFileGroups_.empty()) {
+    VELOX_CHECK(hasSpilled_);
+    spillMerger_ = std::make_unique<SpillMerger>(
+        outputType_, pool(), numSpilledRows_, std::move(spillReadFileGroups_));
+  }
 
-    if (!stream) {
+  if (hasSpilled_) {
+    output_ = spillMerger_->getOutput(outputBatchSize_);
+    if (output_ == nullptr) {
       finished_ = true;
+      return nullptr;
+    }
+    return std::move(output_);
+  }
 
-      // Return nullptr if there is no data.
-      if (outputSize_ == 0) {
-        return nullptr;
-      }
+  // No spill happens.
+  finished_ = lastRun;
```

Comment:
We can always set finished_ to true if we start to read from spilled files?

Reply:
It would disrupt the driver framework, say, call `noMoreInput` of the nextOP?